libs/parser/parser.c

```diff
@@ -0,0 +1,425 @@
+
+	if (p == NULL || path == NULL) {
+		fprintf(stderr, "Parser: invalid arguments. Only non NULL arguments are correct.\n");
+		return -1;
+	}
+
+	file = fopen(path, "r");
+	if (file == NULL) {
+		fprintf(stderr, "Parser: there was an error during opening the file %s\n", path);
+		return -1;
+	}
+
+	while (getline(&line, &lineLen, file) != -1) {
+		switch (parser_lineTypeGet(line)) {
+			case type_Header:
+				/* Invoking converter for the previous header. If this is first header, then `headerInfo` is NULL */
+				if (headerInfo != NULL) {
+					err = headerInfo->converter(p->fieldsMap);
+					hmap_clear(p->fieldsMap);
+					if (err != 0) {
+						fprintf(stderr, "Parser: there was an error during converter invocation\n");
+						break;
+					}
+				}
+				fieldsCnt = 0;
+
+				if (parser_headerGet(line, header) != 0) {
+					fprintf(stderr, "Parser: where was an error during parsing header \"%s\"\n", line);
+					err = -1;
+					break;
+				}
+
+				headerInfo = hmap_get(p->headersMap, header);
+				if (headerInfo == NULL) {
+					fprintf(stderr, "Parser: undefined header was found \"%s\"\n", header);
+					err = -1;
+					break;
+				}
+
+				break;
+
+			case type_Field:
+				/* Checking the case when there is no header above field in file */
+				if (headerInfo == NULL) {
+					fprintf(stderr, "Parser: file is incorrect. All fields have to belong to a header\n\"%s\"\n", line);
+					err = -1;
+					break;
+				}
+
+				if (fieldsCnt >= p->fieldsMap->capacity) {
+					fprintf(stderr, "Parser: maximum number of fields was exceeded\n");
+					err = -1;
+					break;
+				}
+
+				if (parser_fieldGet(line, &(p->fields[fieldsCnt])) != 0) {
+					fprintf(stderr, "Parser: an error occurred during field parsing \"%s\"\n", line);
+					err = -1;
+					break;
+				}
+
+				if (hmap_insert(p->fieldsMap, p->fields[fieldsCnt].field, p->fields[fieldsCnt].value) != 0) {
+					fprintf(stderr, "Parser: incorrect file. Header \"%s\" have multiple fields with name \"%s\"\n", header, p->fields[fieldsCnt].field);
+					err = -1;
+					break;
+				}
+
+				fieldsCnt++;
+				break;
+
+			case type_Comment:
+				break;
+
+			case type_InvalidLine:
+			default:
+				err = -1;
+				break;
+		}
+
+		if (err != 0) {
+			hmap_clear(p->fieldsMap);
+			fclose(file);
+			free(line);
+			return -1;
+		}
+	}
+
+	res = 0;
+
+	if (feof(file) == 0) {
+		fprintf(stderr, "Parser: error occurred during file parsing\n");
+		err = -1;
+	}
+
+	if (res == 0 && headerInfo != NULL) {
+		err = headerInfo->converter(p->fieldsMap);
+
+		if (err != 0) {
+			fprintf(stderr, "Parser: there was an error during converter invocation\n");
+		}
+	}
+
+	hmap_clear(p->fieldsMap);
+	fclose(file);
+	free(line);
+
+	return err;
+}
+
+
+void parser_clear(parser_t *p)
+{
+	if (p == NULL) {
+		return;
+	}
+
+	hmap_clear(p->headersMap);
+}
+
+
+__attribute__((constructor)) static void parser_compileRegex(void)
+{
+	if (regcomp(&(parser_common.commentRegex), COMMENT_PATTERN, REG_EXTENDED) != 0) {
```

Comment:
Please, apply to the rest of the code
```suggestion
	if (regcomp(&parser_common.commentRegex, COMMENT_PATTERN, REG_EXTENDED) != 0) {
```